source/Engine.cpp

```diff
@@ -1464,25 +1477,46 @@ void Engine::CalculateStep()
 	const Ship *flagship = player.Flagship();
 	bool flagshipWasUntargetable = (flagship && !flagship->IsTargetable());
 	bool wasHyperspacing = (flagship && flagship->IsEnteringHyperspace());
+
 	// First, move the player's flagship.
 	if(flagship)
-		MoveShip(player.FlagshipPtr());
+		MoveShip(player.FlagshipPtr(), newVisuals, newFlotsam, newShips, newProjectiles);
 	const System *flagshipSystem = (flagship ? flagship->GetSystem() : nullptr);
 	bool flagshipIsTargetable = (flagship && flagship->IsTargetable());
 	bool flagshipBecameTargetable = flagshipWasUntargetable && flagshipIsTargetable;
+
 	// Then, move the other ships.
-	for(const shared_ptr<Ship> &it : ships)
+	// Keep separate lists for each thread to avoid lock contention.
+	vector<mutex> bufferLocks(thread::hardware_concurrency());
```

Comment:
if this is fixed size, why not use an array?

Reply:
Because I didn't think of it.

Reply:
Actually, I need the size in the method I pass it to.

Reply:
no I mean an std::array, and if you need the size, thread::hardware_concurrency() is constant i think?

Reply:
It is, I just don't want to use it in too many places.

Reply:
> thread::hardware_concurrency() is constant

Not constant enough. The specification doesn't even say it cannot change in runtime, and I'd need it to be a constexpr to use with std::array (because of the template).

Reply:
ok